### PR TITLE
add podmaster support for etcd tls

### PR DESCRIPTION
**Pull request by roboll**
add support for podmaster to consume tls secured etcd.

cc. @brendandburns (since you are listed as the maintainer in the dockerfile for podmaster)

### CODE REVIEW

File: pod-master/podmaster.go [hunk 1]
```diff
@@ -176,7 +186,16 @@ func main() {
 	validateFlags(&c)
 
 	machines := strings.Split(c.etcdServers, ",")
-	etcdClient := etcd.NewClient(machines)
+	var etcdClient *etcd.Client
+	if c.etcdCert != "" && c.etcdKey != "" && c.etcdCA != "" {
```
Comment (brendandburns) [inline pod-master/podmaster.go]: Just test one here, you've already validated that if one is empty they all need to be empty.

File: pod-master/podmaster.go [hunk 2]
```diff
@@ -159,6 +165,10 @@ func validateFlags(c *config) {
 	if len(c.dest) == 0 {
 		glog.Fatalf("--dest-file=<some-file> is required")
 	}
+	//either all are populated, or none are populated
```
Comment (brendandburns) [inline pod-master/podmaster.go]: please move this into the validateFlags function (the ability to return early will clean up this condition)

Comment (brendandburns) [inline pod-master/podmaster.go]: oops, silly git diff made this look like it wasn't in there already...

File: pod-master/podmaster.go [hunk 3]
```diff
@@ -159,6 +165,10 @@ func validateFlags(c *config) {
...
+	if !((c.etcdCert != "" && c.etcdCA != "" && c.etcdKey != "") || (c.etcdCert == "" && c.etcdCA == "" && c.etcdKey == "")) {
```
Comment (brendandburns) [inline pod-master/podmaster.go]: I think that:

```go
if len(c.etcdCert) != 0 || len(c.etcdCA) != 0 || len(c.etcdKey) != 0 {
  if len(c.etcdCert) == 0 || len(c.etcdCA) == 0 || len(c.etcdKey) == 0 {
    glog.Fatalf(...)
  }
}
```

is easier to read.

File: pod-master/podmaster.go [hunk 4]
```diff
@@ -138,6 +141,9 @@ func copyFile(src, dest string) error {
 
 func initFlags(c *config) {
 	pflag.StringVar(&c.etcdServers, "etcd-servers", "", "The comma-seprated list of etcd servers to use")
+	pflag.StringVar(&c.etcdCA, "etcd-ca", "", "The CA file for etcd tls.")
```
Comment (brendandburns) [inline pod-master/podmaster.go]: add these to https://github.com/kubernetes/contrib/blob/master/hack/verify-flags/known-flags.txt

**Comment by brendandburns** [comment]
Thanks for the PR!  small comments.

**Comment by roboll** [comment]
thanks for the comments @brendandburns, everything should be addressed.

**Comment by glerchundi** [comment]
WDYT about using etcd-config in order to be aligned with the apiserver?

**Comment by roboll** [comment]
yeah, i think thats a better idea. good call.

**Comment by glerchundi** [comment]
yep.

It seems like kubernetes core is going to use the new etcd client `github.com/coreos/etcd/client` (not the deprecated one: `github.com/coreos/go-etcd`) which doesn't support reading config from a custom formatted file. Although, in order to be backward compatible maybe that parsing should be made by kubernetes itself...

/ref kubernetes/kubernetes#11962

**Comment by rwehner** [comment]
This PR appears to address Issue https://github.com/kubernetes/contrib/issues/133. Just wanted to make someone reference between the two in case someone (like me) comes across the issue but somehow fails to find the this PR quickly.

I'm looking forward to using this when merged. thanks!

**Comment by glerchundi** [comment]
@rwehner @roboll until this is merged and available for all, i created a public image which uses `etcd-config` parameter as `kube-apiserver` does: `quay.io/saltosystems/podmaster.

**Comment by biwwy** [comment]
I had to do etcd-config file version for my own use, so if anybody wants to use etcd-config from a file, there is a version of pod-master here: https://github.com/biwwy/contrib/commit/79664e46d38cb03d38ca6425e7cc57c0bd2b38e0

**Comment by roboll** [comment]
revisiting this, i poked around all the relevant issues and didn't find any details leading me to believe this was going down the wrong path. 

updating podmaster to use the new etcd config is a little more intense, and probably not as high a priority as this.

@brendandburns is this good for merge?

**Comment by roboll** [comment]
going to close this per https://github.com/kubernetes/contrib/issues/133#issuecomment-163371391

**Pull request closed without merging**